### PR TITLE
Add four Pimoroni RP2350 boards

### PR DESCRIPTION
**Pull request by tannewt**

### CODE REVIEW

**Comment by tannewt** [comment]
cc: @todbot @ZodiusInfuser @Gadgetoid

**Review by dhalbert — CHANGES_REQUESTED** [review]
Hi - the files all need new-style headers.

File: ports/raspberrypi/boards/pimoroni_pico_plus2/pins.c [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    { MP_ROM_QSTR(MP_QSTR_GP15), MP_ROM_PTR(&pin_GPIO15) },
+    { MP_ROM_QSTR(MP_QSTR_GP16), MP_ROM_PTR(&pin_GPIO16) },
+    { MP_ROM_QSTR(MP_QSTR_GP17), MP_ROM_PTR(&pin_GPIO17) },
+    { MP_ROM_QSTR(MP_QSTR_GP18), MP_ROM_PTR(&pin_GPIO18) },
+    { MP_ROM_QSTR(MP_QSTR_GP19), MP_ROM_PTR(&pin_GPIO19) },
+    { MP_ROM_QSTR(MP_QSTR_GP20), MP_ROM_PTR(&pin_GPIO20) },
+    { MP_ROM_QSTR(MP_QSTR_GP21), MP_ROM_PTR(&pin_GPIO21) },
+    { MP_ROM_QSTR(MP_QSTR_GP22), MP_ROM_PTR(&pin_GPIO22) },
+
+    { MP_ROM_QSTR(MP_QSTR_SMPS_MODE), MP_ROM_PTR(&pin_GPIO23) },
```
Comment (ZodiusInfuser) [inline ports/raspberrypi/boards/pimoroni_pico_plus2/pins.c:30]: GPIO23 is not wired up on Pico Plus 2, so these two lines can be removed.

**Comment by ZodiusInfuser** [comment]
Thanks for setting these up @tannewt ! I take it you saw the USB PIDs appear in Pi's repo?

I've had a look over and they overall seem good. There are a few things that could be added, so I'll raise separate comments for those.

File: ports/raspberrypi/boards/pimoroni_pico_plus2/pins.c [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    { MP_ROM_QSTR(MP_QSTR_GP23), MP_ROM_PTR(&pin_GPIO23) },
```
Comment (ZodiusInfuser) [inline ports/raspberrypi/boards/pimoroni_pico_plus2/pins.c:31]: GPIO23 is not wired up on Pico Plus 2, so these two lines can be removed

File: ports/raspberrypi/boards/pimoroni_pico_plus2/pins.c [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    { MP_ROM_QSTR(MP_QSTR_GP26), MP_ROM_PTR(&pin_GPIO26) },
+    { MP_ROM_QSTR(MP_QSTR_A0), MP_ROM_PTR(&pin_GPIO26) },
+
+    { MP_ROM_QSTR(MP_QSTR_GP27_A1), MP_ROM_PTR(&pin_GPIO27) },
+    { MP_ROM_QSTR(MP_QSTR_GP27), MP_ROM_PTR(&pin_GPIO27) },
+    { MP_ROM_QSTR(MP_QSTR_A1), MP_ROM_PTR(&pin_GPIO27) },
+
+    { MP_ROM_QSTR(MP_QSTR_GP28_A2), MP_ROM_PTR(&pin_GPIO28) },
+    { MP_ROM_QSTR(MP_QSTR_GP28), MP_ROM_PTR(&pin_GPIO28) },
+    { MP_ROM_QSTR(MP_QSTR_A2), MP_ROM_PTR(&pin_GPIO28) },
```
Comment (ZodiusInfuser) [inline ports/raspberrypi/boards/pimoroni_pico_plus2/pins.c:49]: GP26, GP27, GP28, do not offer Analog functionality on the B chip.

These should instead just be:
```python
    { MP_ROM_QSTR(MP_QSTR_GP26), MP_ROM_PTR(&pin_GPIO26) },
    { MP_ROM_QSTR(MP_QSTR_GP27), MP_ROM_PTR(&pin_GPIO27) },
    { MP_ROM_QSTR(MP_QSTR_GP28), MP_ROM_PTR(&pin_GPIO28) },
```

For ADC access these additional lines are needed in relevant place:
```python
    { MP_ROM_QSTR(MP_QSTR_GP40_A0), MP_ROM_PTR(&pin_GPIO40) },
    { MP_ROM_QSTR(MP_QSTR_GP40), MP_ROM_PTR(&pin_GPIO40) },
    { MP_ROM_QSTR(MP_QSTR_A0), MP_ROM_PTR(&pin_GPIO40) },

    { MP_ROM_QSTR(MP_QSTR_GP41_A1), MP_ROM_PTR(&pin_GPIO41) },
    { MP_ROM_QSTR(MP_QSTR_GP41), MP_ROM_PTR(&pin_GPIO41) },
    { MP_ROM_QSTR(MP_QSTR_A1), MP_ROM_PTR(&pin_GPIO41) },

    { MP_ROM_QSTR(MP_QSTR_GP42_A2), MP_ROM_PTR(&pin_GPIO42) },
    { MP_ROM_QSTR(MP_QSTR_GP42), MP_ROM_PTR(&pin_GPIO42) },
    { MP_ROM_QSTR(MP_QSTR_A2), MP_ROM_PTR(&pin_GPIO42) },
```

File: ports/raspberrypi/boards/pimoroni_pico_plus2/pins.c [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+    { MP_ROM_QSTR(MP_QSTR_A3), MP_ROM_PTR(&pin_GPIO29) },
+    { MP_ROM_QSTR(MP_QSTR_VOLTAGE_MONITOR), MP_ROM_PTR(&pin_GPIO29) },
```
Comment (ZodiusInfuser) [inline ports/raspberrypi/boards/pimoroni_pico_plus2/pins.c:52]: A3/VOLTAGE_MONITOR is now on GP43.

File: ports/raspberrypi/boards/pimoroni_pico_plus2/pins.c [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+    { MP_ROM_QSTR(MP_QSTR_USER_SW), MP_ROM_PTR(&pin_GPIO45) },
+    { MP_ROM_QSTR(MP_QSTR_BUTTON), MP_ROM_PTR(&pin_GPIO45) },
+
+    { MP_ROM_QSTR(MP_QSTR_STEMMA_I2C), MP_ROM_PTR(&board_i2c_obj) },
+};
```
Comment (ZodiusInfuser) [inline ports/raspberrypi/boards/pimoroni_pico_plus2/pins.c:58]: This board has our new SP/CE connector, so can have these pin definitions adding:

```python
    { MP_ROM_QSTR(MP_QSTR_SPICE_CS), MP_ROM_PTR(&pin_GPIO33) },
    { MP_ROM_QSTR(MP_QSTR_SPICE_SCK), MP_ROM_PTR(&pin_GPIO34) },
    { MP_ROM_QSTR(MP_QSTR_SCK), MP_ROM_PTR(&pin_GPIO34) },
    { MP_ROM_QSTR(MP_QSTR_SPICE_TX), MP_ROM_PTR(&pin_GPIO35) },
    { MP_ROM_QSTR(MP_QSTR_MOSI), MP_ROM_PTR(&pin_GPIO35) },
    { MP_ROM_QSTR(MP_QSTR_SPICE_RX), MP_ROM_PTR(&pin_GPIO32) },
    { MP_ROM_QSTR(MP_QSTR_MISO), MP_ROM_PTR(&pin_GPIO32) },
    { MP_ROM_QSTR(MP_QSTR_SPICE_BL), MP_ROM_PTR(&pin_GPIO36) },
    { MP_ROM_QSTR(MP_QSTR_TFT_BACKLIGHT), MP_ROM_PTR(&pin_GPIO36) },

    { MP_ROM_QSTR(MP_QSTR_SPI), MP_ROM_PTR(&board_spi_obj) },
```

File: ports/raspberrypi/boards/pimoroni_plasma2350/pins.c [hunk 6]
```diff
@@ -0,0 +1,52 @@
+#include "shared-bindings/board/__init__.h"
+
+STATIC const mp_rom_map_elem_t board_module_globals_table[] = {
+    CIRCUITPYTHON_BOARD_DICT_STANDARD_ITEMS
+
+    { MP_ROM_QSTR(MP_QSTR_SW_A), MP_ROM_PTR(&pin_GPIO12) },
+    { MP_ROM_QSTR(MP_QSTR_SW_BOOT), MP_ROM_PTR(&pin_GPIO22) },
```
Comment (ZodiusInfuser) [inline ports/raspberrypi/boards/pimoroni_plasma2350/pins.c:7]: In our schematic we call GPIO22 `USER_SW`. I also see in the other board defs that you add a `BUTTON` def, though those other boards only have the one button.

```python
    { MP_ROM_QSTR(MP_QSTR_USER_SW), MP_ROM_PTR(&pin_GPIO22) },
    { MP_ROM_QSTR(MP_QSTR_BUTTON), MP_ROM_PTR(&pin_GPIO22) },
```

File: ports/raspberrypi/boards/pimoroni_tiny2350/pins.c [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    { MP_ROM_QSTR(MP_QSTR_GP5), MP_ROM_PTR(&pin_GPIO5) },
+    { MP_ROM_QSTR(MP_QSTR_GP6), MP_ROM_PTR(&pin_GPIO6) },
+    { MP_ROM_QSTR(MP_QSTR_GP7), MP_ROM_PTR(&pin_GPIO7) },
+
+    { MP_ROM_QSTR(MP_QSTR_LED_R), MP_ROM_PTR(&pin_GPIO18) },
+    { MP_ROM_QSTR(MP_QSTR_LED_G), MP_ROM_PTR(&pin_GPIO19) },
+    { MP_ROM_QSTR(MP_QSTR_LED_B), MP_ROM_PTR(&pin_GPIO20) },
+
+    { MP_ROM_QSTR(MP_QSTR_SDA), MP_ROM_PTR(&pin_GPIO12) },
+    { MP_ROM_QSTR(MP_QSTR_SCL), MP_ROM_PTR(&pin_GPIO13) },
```
Comment (ZodiusInfuser) [inline ports/raspberrypi/boards/pimoroni_tiny2350/pins.c:20]: These can have their GP pins defined too:

```python
    { MP_ROM_QSTR(MP_QSTR_SDA), MP_ROM_PTR(&pin_GPIO12) },
    { MP_ROM_QSTR(MP_QSTR_GP12), MP_ROM_PTR(&pin_GPIO12) },
    { MP_ROM_QSTR(MP_QSTR_SCL), MP_ROM_PTR(&pin_GPIO13) },
    { MP_ROM_QSTR(MP_QSTR_GP13), MP_ROM_PTR(&pin_GPIO13) },
```

File: ports/raspberrypi/boards/pimoroni_plasma2350/pins.c (hunk 6 above)
Comment (tannewt) [reply ports/raspberrypi/boards/pimoroni_plasma2350/pins.c:7]: Made 12 `BUTTON` and added `BOOT` and `USER_SW` to 22.

**Comment by tannewt** [comment]
All RP2350 headers updated using the script as well as addressed @ZodiusInfuser's comments.

> I take it you saw the USB PIDs appear in Pi's repo?

Yup! I used those.

File: ports/raspberrypi/boards/pimoroni_pico_plus2/pins.c (hunk 3 above)
Comment (ZodiusInfuser) [reply ports/raspberrypi/boards/pimoroni_pico_plus2/pins.c]: @tannewt Thanks for the additions. The above lines for GP40, GP41, GP42 should be included too.

File: ports/raspberrypi/boards/pimoroni_pico_plus2/pins.c [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    { MP_ROM_QSTR(MP_QSTR_SCK), MP_ROM_PTR(&pin_GPIO34) },
+    { MP_ROM_QSTR(MP_QSTR_SPICE_TX), MP_ROM_PTR(&pin_GPIO35) },
+    { MP_ROM_QSTR(MP_QSTR_MOSI), MP_ROM_PTR(&pin_GPIO35) },
+    { MP_ROM_QSTR(MP_QSTR_SPICE_RX), MP_ROM_PTR(&pin_GPIO32) },
+    { MP_ROM_QSTR(MP_QSTR_MISO), MP_ROM_PTR(&pin_GPIO32) },
+    { MP_ROM_QSTR(MP_QSTR_SPICE_BL), MP_ROM_PTR(&pin_GPIO36) },
+    { MP_ROM_QSTR(MP_QSTR_TFT_BACKLIGHT), MP_ROM_PTR(&pin_GPIO36) },
+
+    { MP_ROM_QSTR(MP_QSTR_SPI), MP_ROM_PTR(&board_spi_obj) },
+    { MP_ROM_QSTR(MP_QSTR_STEMMA_I2C), MP_ROM_PTR(&board_i2c_obj) },
```
Comment (ZodiusInfuser) [inline ports/raspberrypi/boards/pimoroni_pico_plus2/pins.c:63]: Missing a regular I2C object: `{ MP_ROM_QSTR(MP_QSTR_I2C), MP_ROM_PTR(&board_i2c_obj) },`

File: ports/raspberrypi/boards/pimoroni_pico_plus2/pins.c [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    { MP_ROM_QSTR(MP_QSTR_VBUS_SENSE), MP_ROM_PTR(&pin_GPIO24) },
+    { MP_ROM_QSTR(MP_QSTR_GP24), MP_ROM_PTR(&pin_GPIO24) },
+
+    { MP_ROM_QSTR(MP_QSTR_LED), MP_ROM_PTR(&pin_GPIO25) },
+    { MP_ROM_QSTR(MP_QSTR_GP25), MP_ROM_PTR(&pin_GPIO25) },
+
+    { MP_ROM_QSTR(MP_QSTR_GP26), MP_ROM_PTR(&pin_GPIO26) },
+    { MP_ROM_QSTR(MP_QSTR_GP27), MP_ROM_PTR(&pin_GPIO27) },
+    { MP_ROM_QSTR(MP_QSTR_GP28), MP_ROM_PTR(&pin_GPIO28) },
+
```
Comment (ZodiusInfuser) [inline ports/raspberrypi/boards/pimoroni_pico_plus2/pins.c:45]: As already mentioned, missing the correct ADC pins:
```python
    { MP_ROM_QSTR(MP_QSTR_GP40_A0), MP_ROM_PTR(&pin_GPIO40) },
    { MP_ROM_QSTR(MP_QSTR_GP40), MP_ROM_PTR(&pin_GPIO40) },
    { MP_ROM_QSTR(MP_QSTR_A0), MP_ROM_PTR(&pin_GPIO40) },

    { MP_ROM_QSTR(MP_QSTR_GP41_A1), MP_ROM_PTR(&pin_GPIO41) },
    { MP_ROM_QSTR(MP_QSTR_GP41), MP_ROM_PTR(&pin_GPIO41) },
    { MP_ROM_QSTR(MP_QSTR_A1), MP_ROM_PTR(&pin_GPIO41) },

    { MP_ROM_QSTR(MP_QSTR_GP42_A2), MP_ROM_PTR(&pin_GPIO42) },
    { MP_ROM_QSTR(MP_QSTR_GP42), MP_ROM_PTR(&pin_GPIO42) },
    { MP_ROM_QSTR(MP_QSTR_A2), MP_ROM_PTR(&pin_GPIO42) },
```

**Comment by tannewt** [comment]
@dhalbert This is ready for a final look.

**Review by dhalbert — APPROVED** [review]
This looks good, and I'm depending on @ZodiusInfuser for the details like VID/PID, etc.

**Comment by tannewt** [comment]
For the record. USB PIDs are publicly documented here: https://github.com/raspberrypi/usb-pid/blob/96bdf45629c09004f2d628b9573d7a901a64eb23/Readme.md?plain=1#L187-L190

**Pull request merged**